consensus/core/src/leader_scoring.rs

```diff
@@ -75,9 +176,197 @@ impl ReputationScores {
+    // that the CommittedSubDag instances are contiguous in commit index order.
+    // Therefore we can guarnatee the blocks of UnscoredSubdag are also sorted
+    // via the commit index.
+    pub blocks: BTreeMap<BlockRef, VerifiedBlock>,
+    pub commit_range: CommitRange,
+}
+
+impl UnscoredSubdag {
+    pub(crate) fn new(context: Arc<Context>, subdags: &[CommittedSubDag]) -> Self {
+        let blocks = subdags
+            .iter()
+            .enumerate()
+            .flat_map(|(subdag_index, subdag)| {
+                if subdag_index == 0 {
+                    subdag.blocks.iter()
+                } else {
+                    let previous_subdag = &subdags[subdag_index - 1];
+                    let expected_next_subdag_index = previous_subdag.commit_index + 1;
+                    assert_eq!(
+                        subdag.commit_index, expected_next_subdag_index,
+                        "Non-contiguous commit index (expected: {}, found: {})",
+                        expected_next_subdag_index, subdag.commit_index
+                    );
+                    subdag.blocks.iter()
+                }
+            })
+            .map(|block| (block.reference(), block.clone()))
+            .collect::<BTreeMap<_, _>>();
+
+        // Guaranteed to have a contiguous list of commit indices
+        let commit_range = CommitRange::new(
+            subdags.first().unwrap().commit_index..subdags.last().unwrap().commit_index,
+        );
+
+        assert!(
+            !blocks.is_empty(),
+            "Attempted to create UnscoredSubdag with no blocks"
+        );
+
+        Self {
+            context,
+            blocks,
+            commit_range,
+        }
+    }
+
+    pub(crate) fn get_leader_round_range(&self) -> (Round, Round) {
+        // Skip genesis round as we don't produce leaders for that round.
+        let first_round = self
+            .blocks
+            .keys()
+            .find(|block_ref| block_ref.round != 0)
+            .map(|block_ref| block_ref.round)
+            .expect("There should be a non-zero round in the set of blocks");
+        let last_round = self.blocks.keys().last().unwrap().round;
+        (first_round, last_round)
+    }
+
+    pub(crate) fn find_supported_block(
+        &self,
+        leader_slot: Slot,
+        from: &VerifiedBlock,
+    ) -> Option<BlockRef> {
+        if from.round() < leader_slot.round {
+            return None;
+        }
+        for ancestor in from.ancestors() {
+            if Slot::from(*ancestor) == leader_slot {
+                return Some(*ancestor);
+            }
+            // Weak links may point to blocks with lower round numbers than strong links.
+            if ancestor.round <= leader_slot.round {
+                continue;
+            }
+            if let Some(ancestor) = self.get_block(ancestor) {
+                if let Some(support) = self.find_supported_block(leader_slot, &ancestor) {
```

Comment:
I would try to avoid recursion if possible. Otherwise the semantic when there are equivocations can be very subtle. Maybe add a TODO.